src/librustc/mir/transform.rs

```diff
@@ -8,9 +8,79 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-use mir::repr::Mir;
+use mir::repr::{Mir, BasicBlockData, BasicBlock};
+use mir::mir_map::MirMap;
 use middle::ty::ctxt;
 
-pub trait MirPass {
-    fn run_on_mir<'tcx>(&mut self, mir: &mut Mir<'tcx>, tcx: &ctxt<'tcx>);
+/// Contains various metadata about the pass.
+pub trait Pass {
+    /// Ordering of the pass. Lower value runs the pass earlier.
+    fn priority(&self) -> usize;
+    // Possibly also `fn name()` and `fn should_run(Session)` etc.
+}
+
+/// Pass which inspects the whole MirMap.
+pub trait MirMapPass: Pass {
+    fn run_pass<'tcx>(&mut self, tcx: &ctxt<'tcx>, map: &mut MirMap<'tcx>);
+}
+
+/// Pass which only inspects MIR of distinct functions.
+pub trait MirPass: Pass {
+    fn run_pass<'tcx>(&mut self, tcx: &ctxt<'tcx>, mir: &mut Mir<'tcx>);
+}
+
+/// Pass which only inspects basic blocks in MIR.
+///
+/// Invariant: The blocks are considered to be fully self-contained for the purposes of this pass –
+/// the pass may not change the list of successors of the block or apply any transformations to
+/// blocks based on the information collected during earlier runs of the pass.
```

Comment:
The point of this invariant is that the pass *should not* store such information. Primarily because I expect these passes to be pipelined, parallelized, etc in the future. The pass must not rely on getting blocks in any deterministic order. (I realise I also forgot to add a similar invariant for the `MirPass`)